addons/mail/tests/discuss/test_discuss_mail_presence.py

```diff
@@ -49,52 +49,40 @@ def _receive_presence(self, sender, recipient):
-        channel._add_members(guests=guest, users=bob)
-        # Now that they share a channel, guest should receive users's presence.
-        self._receive_presence(sender=bob, recipient=guest)
-
-        other_guest = self.env["mail.guest"].create({"name": "OtherGuest"})
-        # Guest should not receive guest's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=other_guest, recipient=guest)
-        channel._add_members(guests=other_guest)
-        # Now that they share a channel, guest should receive guest's presence.
-        self._receive_presence(sender=other_guest, recipient=guest)
-        self.assertEqual(other_guest.im_status, "online")
-
-    def test_receive_presences_as_portal(self):
-        portal = new_test_user(self.env, login="portal_user", groups="base.group_portal")
-        bob = new_test_user(self.env, login="bob_user", groups="base.group_user")
-        # Portal should not receive users's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=bob, recipient=portal)
-        channel = self.env["discuss.channel"]._create_channel(group_id=None, name="General")
-        channel._add_members(users=portal | bob)
-        # Now that they share a channel, portal should receive users's presence.
-        self._receive_presence(sender=bob, recipient=portal)
-
+    def test_presence_access(self):
+        internal_user = new_test_user(self.env, login="internal_user", groups="base.group_user")
+        other_internal_user = new_test_user(
+            self.env, login="other_internal_user", groups="base.group_user"
+        )
+        portal_user = new_test_user(self.env, login="portal_user", groups="base.group_portal")
         guest = self.env["mail.guest"].create({"name": "Guest"})
-        # Portal should not receive guest's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=guest, recipient=portal)
-        channel._add_members(guests=guest)
-        # Now that they share a channel, portal should receive guest's presence.
-        self._receive_presence(sender=guest, recipient=portal)
-        self.assertEqual(guest.im_status, "online")
+        test_cases = [
+            # requested_by - target - has_token - allowed
+            (guest, internal_user, False, False),
+            (guest, internal_user, True, True),
+            (guest, portal_user, False, False),
+            (guest, portal_user, True, True),
+            (internal_user, guest, False, True),
+            (internal_user, guest, True, True),
+            (internal_user, other_internal_user, False, True),
+            (internal_user, other_internal_user, True, True),
+            (internal_user, portal_user, False, True),
+            (internal_user, portal_user, True, True),
```

Comment:
I'm not sure if it helps, but all of those for internal_user could be a single itertool product with True at the end.
And then the rest can be 2 other product: one which has False False, and one which has True True.

This would make the test 3 lines, that are in direct relation to what we want functionally:
internal user -> True
all other cases -> token -> True
all other cases -> no token -> False

Reply:
It could, but frankly I don't think this is easier to read:

```py
list(product(["internal_user"], ["guest", "other_internal", "portal"], [False, True], [True]))
```